src/absil/ilwritepdb.fs

```diff
@@ -0,0 +1,634 @@
+#endif
+
+open Microsoft.FSharp.Compiler.DiagnosticMessage
+open Microsoft.FSharp.Compiler.ErrorLogger
+open Microsoft.FSharp.Compiler.Range
+
+open System.Collections.Generic 
+open System.IO
+
+// -------------------------------------------------------------------- 
+// PDB types
+// --------------------------------------------------------------------  
+type PdbDocumentData = ILSourceDocument
+
+type PdbLocalVar = 
+    { Name: string
+      Signature: byte[] 
+      /// the local index the name corresponds to
+      Index: int32  }
+
+type PdbMethodScope = 
+    { Children: PdbMethodScope array
+      StartOffset: int
+      EndOffset: int
+      Locals: PdbLocalVar array
+      (* REVIEW open_namespaces: pdb_namespace array *) }
+
+type PdbSourceLoc = 
+    { Document: int
+      Line: int
+      Column: int }
+      
+type PdbSequencePoint = 
+    { Document: int
+      Offset: int
+      Line: int
+      Column: int
+      EndLine: int
+      EndColumn: int }
+    override x.ToString() = sprintf "(%d,%d)-(%d,%d)" x.Line x.Column x.EndLine x.EndColumn
+
+type PdbMethodData = 
+    { MethToken: int32
+      MethName:string
+      LocalSignatureToken: int32
+      Params: PdbLocalVar array
+      RootScope: PdbMethodScope
+      Range: (PdbSourceLoc * PdbSourceLoc) option
+      SequencePoints: PdbSequencePoint array }
+
+module SequencePoint = 
+    let orderBySource sp1 sp2 = 
+        let c1 = compare sp1.Document sp2.Document
+        if c1 <> 0 then c1 else 
+        let c1 = compare sp1.Line sp2.Line
+        if c1 <> 0 then c1 else 
+        compare sp1.Column sp2.Column 
+        
+    let orderByOffset sp1 sp2 = 
+        compare sp1.Offset sp2.Offset 
+
+/// 28 is the size of the IMAGE_DEBUG_DIRECTORY in ntimage.h 
+let sizeof_IMAGE_DEBUG_DIRECTORY = 28 
+
+[<NoEquality; NoComparison>]
+type PdbData = 
+    { EntryPoint: int32 option
+      Timestamp: int32
+      ModuleID: byte[]
+      Documents: PdbDocumentData[]
+      Methods: PdbMethodData[] 
+      TableRowCounts: int[] }
+
+//---------------------------------------------------------------------
+// Portable PDB Writer
+//---------------------------------------------------------------------
+open System
+open System.Reflection
+open System.Reflection.Metadata
+open System.Reflection.Metadata.Ecma335
+open System.Reflection.Metadata.Ecma335.Blobs
+open System.Reflection.PortableExecutable
+open System.Collections.Generic
+open System.Collections.Immutable
+
+type idd =
+    { iddCharacteristics: int32;
+      iddMajorVersion: int32; (* actually u16 in IMAGE_DEBUG_DIRECTORY *)
+      iddMinorVersion: int32; (* actually u16 in IMAGE_DEBUG_DIRECTORY *)
+      iddType: int32;
+      iddData: byte[];}
+
+let magicNumber = 0x53445352L
+let pdbGetDebugInfo (mvid:byte[]) (filepath:string) = 
+    let iddDataBuffer = 
+        let path = (System.Text.Encoding.UTF8.GetBytes filepath)
+        let buffer = Array.zeroCreate (sizeof<int32> + mvid.Length + sizeof<int32> + path.Length + 1)
```

Comment:
I'm sure this is common sight in compilers and quite efficient / low allocation code, but I'll be less likely to introduce bugs if I was using a buffer writer type of thing rather than keeping track of offset, putting next element and asserting buffer was written till it's length.

if this is meant to stay low level / no buffer writer, why not put an `assert ( buffer.Length = offset + size)` before returning buffer?

At least, since the buffer is an array, there won't be silent writing after it's bound, but  a